src/Auth.ts

```diff
@@ -713,34 +740,86 @@ export class Auth {
     return resource;
   }
 
-  private async getServiceConnectionInfo<TConn>(): Promise<TConn> {
-    const tokenStorage = this.getTokenStorage();
+  private async getConnectionInfoFromStorage<TConn>(): Promise<TConn> {
+    const tokenStorage = this.getConnectionStorage();
     const json: string = await tokenStorage.get();
     return JSON.parse(json);
   }
 
-  public storeConnectionInfo(): Promise<void> {
-    const tokenStorage = this.getTokenStorage();
-    return tokenStorage.set(JSON.stringify(this.service));
+  public async storeConnectionInfo(): Promise<void> {
+    const connectionStorage = this.getConnectionStorage();
+    await connectionStorage.set(JSON.stringify(this.connection));
+
+    let allConnections = await this.getAllConnections();
+
+    if (this.connection.active) {
+      allConnections = allConnections.filter(c => c.identityId !== this.connection.identityId);
+      allConnections.forEach(c => c.active = false);
+      allConnections = [{ ...this.connection as any }, ...allConnections];
+    }
+
+    this._allConnections = allConnections;
+    const allConnectionsStorage = this.getAllConnectionsStorage();
+    await allConnectionsStorage.set(JSON.stringify(allConnections));
   }
 
   public async clearConnectionInfo(): Promise<void> {
-    const tokenStorage = this.getTokenStorage();
-    await tokenStorage.remove();
+    const connectionStorage = this.getConnectionStorage();
+    const allConnectionsStorage = this.getAllConnectionsStorage();
+
+    await connectionStorage.remove();
+    await allConnectionsStorage.remove();
+
     // we need to manually clear MSAL cache, because MSAL doesn't have support
     // for logging out when using cert-based auth
     const msalCache = this.getMsalCacheStorage();
     await msalCache.remove();
   }
 
-  public getTokenStorage(): TokenStorage {
+  public async removeConnectionInfo(logger: Logger, debug: boolean, connection: Connection): Promise<void> {
+    const allConnections = await this.getAllConnections();
+    const isCurrentConnection = this.connection.name === connection.name;
+
+    this._allConnections = allConnections.filter(c => c.name !== connection.name);
+
+    // When using an application identity, there is no account in the MSAL TokenCache
+    if (this.connection.authType !== AuthType.Certificate
+      && this.connection.authType !== AuthType.Secret
+      && this.connection.authType !== AuthType.Identity) {
+      this.clientApplication = await this.getPublicClient(logger, debug);
+
+      if (this.clientApplication) {
+        const tokenCache = this.clientApplication.getTokenCache();
+        const account = await tokenCache.getAccountByLocalId(connection!.identityId!);
```

Comment:
I've removed one !. 
What's your suggestion for the last one? `identityId` is nullable. I'm expecting it to be filled here, but I dislike having to do a null-check at every fricking place where you use such a thing.

how would you advise to solve it?

Reply:
If it's nullable, why are you expecting it to be filled here? Why is that assumption implicit? It's the whole point of using types to isolate yourself from implicit assumptions. If something is optional, we should check for it before using it. If the value is always present, then we shouldn't make it optional.

Reply:
I'm expecting it to be filled, because what we're doing here is removing a connection. We wouldn't be here if there is no connection, a connection always has an identityId value after signing in. Hence it's logical to assume identityId is filled.

It would be better if the identityId property of the connection object would not be nullable. But that would mean we'd need to refactor the code again. The original service object/interface setup and auth flow was created in a less strict typescript era I think 😆

Reply:
I've now worked around it as you can see in my last commit. As I'm saying: this is not exactly 100% ideal. But I do not wish to refactor the auth.connection object setup further.